geniza/annotations/signals.py

```diff
@@ -1,20 +1,34 @@
+import logging
+
 from django.conf import settings
 from django.contrib.admin.models import ADDITION, DELETION, LogEntry
 from django.contrib.auth.models import User
 from django.contrib.contenttypes.models import ContentType
+from django.db.models.signals import post_delete, post_save
 
 from geniza.annotations.models import Annotation
 from geniza.corpus.models import Document
 from geniza.footnotes.models import Footnote, Source
 
+logger = logging.getLogger(__name__)
+
 
 def create_or_delete_footnote(instance, **kwargs):
     """On Annotation save, create digital edition Footnote if necessary.
     On Annotation delete, if it's the only Annotation for the corresponding Footnote,
     delete the Footnote."""
     # get manifest uri from annotation content and match it to a Document
-    manifest_uri = instance.content["target"]["source"]["partOf"]["id"]
-    source_uri = instance.content["dc:source"]
+
+    try:
+        manifest_uri = instance.target_source_manifest_id
+        source_uri = instance.content["dc:source"]
+    except KeyError:
+        # annotations created from tahqiq should have these;
+        # if not present, bail out
+        # (likely only happening in unit tests)
+        return
+
+    # TODO: can we get ids without loading source/doc from db?
```

Comment:
Good point, at least for the document you can do `resolve(urlparse(uri).path).kwargs["pk"]`.

Reply:
yes — was thinking it might be worth separating out the method to get the object from the logic to get the id